tests/core/vm/test_sharding_vm.py

```diff
@@ -7,26 +12,117 @@
+        simple_transfer_contract_address,
+        b'',
+        0,
+        b'',
+        b'',
+        simple_transfer_contract_bytecode
+    )
 
     vm = chain.get_vm()
-    computation = vm.apply_transaction(deploy_tx)
+    computation = vm.apply_transaction(first_deploy_tx)
     assert not computation.is_error
+    gas_used = vm.block.header.gas_used
+    assert gas_used > first_deploy_tx.intrinsic_gas
+    last_gas_used = gas_used
 
     # Transfer ether to recipient
     recipient = decode_hex('0xa94f5374fce5edbc8e2a8697c15331677e6ebf0c')
     amount = 100
-    tx_initiator = contract_address
+    tx_initiator = simple_transfer_contract_address
     transfer_tx = new_sharding_transaction(tx_initiator, recipient, amount, b'', b'', b'')
 
-    computation = vm.execute_transaction(transfer_tx)
+    computation = vm.apply_transaction(transfer_tx)
     assert not computation.is_error
+    gas_used = vm.block.header.gas_used - last_gas_used
+    assert gas_used > transfer_tx.intrinsic_gas
+    last_gas_used = vm.block.header.gas_used
     with vm.state.state_db(read_only=True) as state_db:
         assert state_db.get_balance(recipient) == amount
+
+    # Second test: contract that deploy new contract with CREATE2
+    second_deploy_tx = new_sharding_transaction(
+        CREATE2_contract_address,
+        b'',
+        0,
+        b'',
+        b'',
+        CREATE2_contract_bytecode
+    )
+
+    computation = vm.apply_transaction(second_deploy_tx)
+    assert not computation.is_error
+    gas_used = vm.block.header.gas_used - last_gas_used
+    assert gas_used > second_deploy_tx.intrinsic_gas
+    last_gas_used = vm.block.header.gas_used
+
+    # Invoke the contract to deploy new contract
+    tx_initiator = CREATE2_contract_address
+    invoke_tx = new_sharding_transaction(tx_initiator, b'', 0, b'', b'', b'')
+
+    computation = vm.apply_transaction(invoke_tx)
+    assert not computation.is_error
+    gas_used = vm.block.header.gas_used - last_gas_used
+    assert gas_used > invoke_tx.intrinsic_gas
+    with vm.state.state_db(read_only=True) as state_db:
+        newly_deployed_contract_address = generate_CREATE2_contract_address(
+            int_to_big_endian(0),
+            simple_contract_factory_bytecode
+        )
+        assert state_db.get_code(newly_deployed_contract_address) == b'\xbe\xef'
+        assert state_db.get_storage(CREATE2_contract_address, 0) == 1
+
+
+def test_CREATE2_deploy_contract_edge_cases(shard_chain_without_block_validation):  # noqa: F811
+    # First case: computed contract address not the same as provided in `transaction.to`
+    chain = shard_chain_without_block_validation
+    first_failed_deploy_tx = new_sharding_transaction(
+        simple_transfer_contract_address,
+        b'',
+        0,
+        b'',
+        b'',
+        '0xf3',
+    )
+
+    vm = chain.get_vm()
+    computation = vm.apply_transaction(first_failed_deploy_tx)
+    assert computation.is_error
```

Comment:
Maybe add a check for `isinstance(computation._error, IncorrectContractCreationAddress)`